akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -0,0 +1,296 @@
+import scala.concurrent.duration.{ Deadline, FiniteDuration }
+
+/**
+ * Used to start the termination process of an Akka HTTP server.
+ *
+ * TODO more details here?
+ */
+// "Hasta la vista, baby."
+@DoNotInherit
+abstract class ServerTerminator {
+  /**
+   * Initiate the termination sequence of this server.
+   *
+   * TODO more docs here
+   */
+  def terminate(deadline: FiniteDuration)(implicit ex: ExecutionContext): Future[HttpTerminated]
+}
+
+/** INTERNAL API */
+@InternalApi
+object MasterServerTerminator {
+  sealed trait State
+  final case class AliveConnectionTerminators(ts: Set[ServerTerminator]) extends State
+  final case class Terminating(deadline: Deadline) extends State
+}
+
+/** INTERNAL API: Collects signals from per-connection terminators and manages the termination process kickoff */
+@InternalApi
+final class MasterServerTerminator(log: LoggingAdapter) extends ServerTerminator {
+  import MasterServerTerminator._
+
+  // if we get a termination signal from the outer (user-land) it has to send this signal to all existing connection
+  // terminators such that they can initiate the termination (draining, failing) of their respective connections.
+
+  private val terminators = new AtomicReference[MasterServerTerminator.State](MasterServerTerminator.AliveConnectionTerminators(Set.empty))
+  private val termination = Promise[HttpTerminated]()
+
+  /**
+   * Registers an per-connection terminator.
+   * Once the master terminator gets the termination signal,
+   * it will delegate the signal to all registered terminators (existing connections).
+   *
+   * @return true if registered successfully and not terminating, false if termination in-flight
+   *         (and the terminators `terminate()` will be invoked in that case automatically)
+   */
+  @tailrec def registerConnection(terminator: ServerTerminator)(implicit ec: ExecutionContext): Boolean = {
+    terminators.get() match {
+      case v @ AliveConnectionTerminators(ts) ⇒
+        terminators.compareAndSet(v, v.copy(ts = ts + terminator)) ||
+          registerConnection(terminator) // retry
+
+      case Terminating(deadline) ⇒
+        terminator.terminate(deadline.timeLeft)(ec)
+        false // termination is in progress already, we did not register but immediately issue termination
+    }
+  }
+
+  // If a connection attempts to register once termination has started, it will immediately be rejected (though
+  // since termination also implies unbinding such new connections should not really happen).
+  def terminate(timeout: FiniteDuration)(implicit ex: ExecutionContext): Future[HttpTerminated] = {
+    terminators.get() match {
+      case v @ AliveConnectionTerminators(emptyTs) if emptyTs.isEmpty ⇒
+        if (terminators.compareAndSet(v, Terminating(timeout.fromNow))) {
+          // no connections exist to be terminated, and if some arrive after us, they will see the Terminating and terminate
+          termination.success(HttpServerTerminated)
```

Comment:
sure that it's safe? relation to below `termination.completeWith`? otherwise `trySuccess`

Reply:
Yeah I think it's safe due to the CAS protecting it, but won't hurt to make it a trySuccess I think